### PR TITLE
adding dataproc cluster start and stop operators

### PR DESCRIPTION
**Pull request by arezamoosavi**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 1]
```diff
@@ -659,6 +680,264 @@ def execute(self, context: 'Context') -> dict:
         return Cluster.to_dict(cluster)
 
 
+class DataprocStartClusterOperator(BaseOperator):
```
Comment (lwyszomi) [inline airflow/providers/google/cloud/operators/dataproc.py]: Could you add the link the the Dataproc cluster. Check 'operator_extra_links'

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 2]
```diff
@@ -659,6 +680,264 @@ def execute(self, context: 'Context') -> dict:
...
+        self._start_cluster(hook)
+        cluster = self._get_cluster(hook)
+        self._handle_error_state(hook, cluster)
+        if cluster.status.state == cluster.status.State.STARTING:
+            # Wait for cluster to be running
+            cluster = self._wait_for_cluster_in_starting_state(hook)
+            self._handle_error_state(hook, cluster)
+
+
+class DataprocStopClusterOperator(BaseOperator):
```
Comment (lwyszomi) [inline airflow/providers/google/cloud/operators/dataproc.py]: same for this operator.

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 3]
```diff
@@ -659,6 +680,264 @@ def execute(self, context: 'Context') -> dict:
...
+        # Save data required to display extra link no matter what the cluster status will be
+        self.xcom_push(
+            context,
+            key="cluster_conf",
+            value={
+                "cluster_name": self.cluster_name,
+                "region": self.region,
+                "project_id": self.project_id,
+            },
+        )
```
Comment (lwyszomi) [inline airflow/providers/google/cloud/operators/dataproc.py]: please use new approach for the links: more info in the PR: https://github.com/apache/airflow/pull/21267

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 4]
```diff
@@ -659,6 +680,264 @@ def execute(self, context: 'Context') -> dict:
...
```
Comment (lwyszomi) [inline airflow/providers/google/cloud/operators/dataproc.py]: same here

**Review by lwyszomi — COMMENTED** [review]
Generally look good but could you fix styling in your code.

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 1 above)
Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: It has been fixed.

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 2 above)
Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: the same here.

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 3 above)
Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: Here I am going to push and not pull; so they are not the same, I guess.

Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: Also, there is another way for set: XCom.set. In the code, they all used xcom_push in the execution section.

Comment (lwyszomi) [reply airflow/providers/google/cloud/operators/dataproc.py]: I mean a @turbaszek comment to create or use static `persist` method and avoid copy/paste whole push logic.

Example in the Metastrore operators: https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/operators/dataproc_metastore.py#L58

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 5]
```diff
@@ -659,6 +680,268 @@ def execute(self, context: 'Context') -> dict:
...
+                raise AirflowException(
+                    f"Cluster {self.cluster_name} is still CREATING state, aborting")
+            time.sleep(time_to_sleep)
+            time_left = time_left - time_to_sleep
+            cluster = self._get_cluster(hook)
+        return cluster
+
+    def execute(self, context: 'Context') -> None:
+        self.log.info('Starting cluster: %s', self.cluster_name)
+        hook = DataprocHook(gcp_conn_id=self.gcp_conn_id,
```
Comment (josh-fell) [inline airflow/providers/google/cloud/operators/dataproc.py]: You could simply the implementation in the operator by having `hook` as an instance attribute of the operator. Using the instance attribute, you wouldn't have to pass the object around and the other methods like `_get_cluster()` and `_start_cluster()` could be replaced by simply calling `self.hook.get_cluster()` and `self.hook.start_cluster()` directly.

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 6]
```diff
@@ -659,6 +680,268 @@ def execute(self, context: 'Context') -> dict:
...
+                raise AirflowException(
+                    f"Cluster {self.cluster_name} is still STOPPING state, aborting")
+            time.sleep(time_to_sleep)
+            time_left = time_left - time_to_sleep
+            cluster = self._get_cluster(hook)
+        return cluster
+
+    def execute(self, context: 'Context') -> None:
+        self.log.info('Stopping cluster: %s', self.cluster_name)
+        hook = DataprocHook(gcp_conn_id=self.gcp_conn_id,
```
Comment (josh-fell) [inline airflow/providers/google/cloud/operators/dataproc.py]: Same comment here re: using an instance attribute for the hook object.

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 7]
```diff
@@ -698,7 +981,8 @@ class DataprocScaleClusterOperator(BaseOperator):
         account from the list granting this role to the originating account (templated).
     """
 
-    template_fields: Sequence[str] = ('cluster_name', 'project_id', 'region', 'impersonation_chain')
+    template_fields: Sequence[str] = (
+        'cluster_name', 'project_id', 'region', 'impersonation_chain')
```
Comment (josh-fell) [inline airflow/providers/google/cloud/operators/dataproc.py]: This (and other) formatting changes look unrelated, and the static checks in the CI probably won't like them. You could run these static checks locally though. Check out [this guide](https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst) for some pointers.

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 3 above)
Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: Hi I rebase the updates on dataproc and create this [PR](https://github.com/apache/airflow/pull/21913)

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 2 above)
Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: I rebased the updates on dataproc and create this [PR](https://github.com/apache/airflow/pull/21913)

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 4 above)
Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: it has been fixed in another PR so I rebased the updates on dataproc and create this [PR](https://github.com/apache/airflow/pull/21913)

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 5 above)
Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: done, I rebased the updates on dataproc and create this [PR](https://github.com/apache/airflow/pull/21913)

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 7 above)
Comment (arezamoosavi) [reply airflow/providers/google/cloud/operators/dataproc.py]: done, thanks for the feedback.

**Pull request closed without merging**